src/servicesHost.ts

```diff
@@ -509,3 +520,45 @@ function populateDependencyGraphs(
     ] = true;
   });
 }
+
+function addCache(servicesHost: typescript.ModuleResolutionHost) {
+  const clearCacheFuncs: (() => void)[] = [];
+  if (servicesHost.fileExists !== undefined) {
+    const cache = createCache(servicesHost.fileExists);
+    servicesHost.fileExists = cache.cached;
+    clearCacheFuncs.push(cache.clear);
+  }
+
+  if (servicesHost.directoryExists !== undefined) {
+    const cache = createCache(servicesHost.directoryExists);
+    servicesHost.directoryExists = cache.cached;
+    clearCacheFuncs.push(cache.clear);
+  }
+
+  if (servicesHost.realpath !== undefined) {
+    const cache = createCache(servicesHost.realpath);
+    servicesHost.realpath = cache.cached;
+    clearCacheFuncs.push(cache.clear);
+  }
+
+  return () => clearCacheFuncs.forEach(clear => clear());
+}
+
```

Comment:
I wonder if we could reduce some duplication here?

```
['fileExists', 'directoryExists', 'realpath'].forEach(propertyToCache => {
      if (servicesHost[propertyToCache] !== undefined) {
        const cache = createCache(servicesHost[propertyToCache]);
        servicesHost[propertyToCache] = cache.cached;
        clearCacheFuncs.push(cache.clear);
      }
});
```

Reply:
Yeah it's possible you might need to do this:

```
['fileExists', 'directoryExists', 'realpath'].forEach((propertyToCache: 'fileExists' | 'directoryExists' | 'realpath') => {
      if (servicesHost[propertyToCache] !== undefined) {
        const cache = createCache(servicesHost[propertyToCache]);
        servicesHost[propertyToCache] = cache.cached;
        clearCacheFuncs.push(cache.clear);
      }
});
```

Reply:
I bet you could equally use `Proxy` for this functionality too: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy#Basic_example

Never actually used a `Proxy` in JS yet 😄

Reply:
Ah node 6 minimum... https://node.green/#ES2015-built-ins-Proxy

Reply:
It seems that it wouldn't work because TypeScript infer type `(((arg: string) => string) | (arg: string) => boolean))` to type `(arg: string) => string | boolean`, but they are not the same types.

Reply:
Even if I change to this:

```ts
  [
    'fileExists' as 'fileExists',
    'directoryExists' as 'directoryExists',
    'realpath' as 'realpath',
  ].forEach(propertyToCache => {
    const origFn = servicesHost[propertyToCache];
    if (origFn !== undefined) {
      const cache = createCache(origFn);
      servicesHost[propertyToCache] = cache.cached;
      clearCacheFuncs.push(cache.clear);
    }
  });
```

and `createCache` to `function createCache<TFn extends (arg: string) => TOut, TOut = ReturnType<TFn>>(func: TFn) {` TypeScript fails with error ` Type '(arg: string) => string | boolean' is not assignable to type '((fileName: string) => boolean) | ((path: string) => string) | undefined'.`.

I can hack it with casting via `as`, but I'm not sure that it is good idea 🙂

Reply:
It seems the following code compiles well, but it might be a little bit tricky:

```ts
  [
    'fileExists' as 'fileExists',
    'directoryExists' as 'directoryExists',
    'realpath' as 'realpath',
  ].forEach(propertyToCache => {
    const origFn = servicesHost[propertyToCache];
    if (origFn !== undefined) {
      const cache = createCache<ReturnType<typeof origFn>>(origFn);
      servicesHost[propertyToCache] = cache.cached as typescript.ModuleResolutionHost[typeof propertyToCache];
      clearCacheFuncs.push(cache.clear);
    }
  });
```

What do you think?

Reply:
I can avoid specifying `<ReturnType<typeof origFn>>`  by changing `createCache`'s declaration as in https://github.com/TypeStrong/ts-loader/pull/829#discussion_r214755099 also.

Reply:
To add additional compile-time checks of property names we can do the following:

```ts
  type CachedProperty = Extract<
    keyof typescript.ModuleResolutionHost,
    'fileExists' | 'directoryExists' | 'realpath'
  >;
  const cachedProperties: CachedProperty[] = [
    'fileExists',
    'directoryExists',
    'realpath'
  ];

  cachedProperties.forEach((propertyToCache: CachedProperty) => {
    const origFn = servicesHost[propertyToCache];
    if (origFn !== undefined) {
      const cache = createCache<ReturnType<typeof origFn>>(origFn);
      servicesHost[
        propertyToCache
      ] = cache.cached as typescript.ModuleResolutionHost[CachedProperty];
      clearCacheFuncs.push(cache.clear);
    }
  });
```